packages/core/src/saml-applications/routes/utils.ts

```diff
@@ -0,0 +1,299 @@
+import { parseJson } from '@logto/connector-kit';
+import { generateStandardId } from '@logto/shared';
+import { tryThat } from '@silverhand/essentials';
+import camelcaseKeys from 'camelcase-keys';
+import { got } from 'got';
+import saml from 'samlify';
+import { ZodError, z } from 'zod';
+
+import RequestError from '#src/errors/RequestError/index.js';
+import { fetchOidcConfigRaw, getRawUserInfoResponse } from '#src/sso/OidcConnector/utils.js';
+import { idTokenProfileStandardClaimsGuard } from '#src/sso/types/oidc.js';
+import { oidcTokenResponseGuard, type IdTokenProfileStandardClaims } from '#src/sso/types/oidc.js';
+import assertThat from '#src/utils/assert-that.js';
+
+import {
+  samlLogInResponseTemplate,
+  samlAttributeNameFormatBasic,
+  samlValueXmlnsXsi,
+} from '../libraries/consts.js';
+
+// We only support email and persistent format at the moment.
+const getSamlNameId = (user: IdTokenProfileStandardClaims, idpNameIDFormat?: string | string[]) => {
+  // If IdP has specified nameIDFormat, use it
```

Comment:
Side topic: To support different nameID mapping and format requirements, should we add it as a SAML IdP config?  Use logto userId by default if not provided.

Reply:
Yes, we need to add a configurable field. This PR is the same code that we used to demo IdP-init SSO flow, and they need email address for name ID for now. Before we add the config, we keep using email address by default for now.